File: basic-service-invocation/microservices/svc/blog/handler/blog.go
package handler

import (
	"context"

	pb "github.com/kzmake/examples/basic-service-invocation/gen/go/blog/v1"
)

type blog struct {
	pb.UnimplementedBlogServiceServer
}

var _ pb.BlogServiceServer = new(blog)

func NewBlog() pb.BlogServiceServer {
	return &blog{}
}

func (h *blog) Create(ctx context.Context, req *pb.CreateRequest) (*pb.CreateResponse, error) {
	blog := &pb.Blog{
		Url: "https://example.com/" + req.GetUsername(),
	}

	// do something

	return &pb.CreateResponse{Blog: blog}, nil
}

File: basic-service-invocation-py/microservices/svc/blog/handler/blog.py
from basic.blog.v1.blog_grpc import BlogServiceBase
from basic.blog.v1.blog_pb2 import Blog, CreateRequest, CreateResponse
from grpclib.server import Stream


class BlogService(BlogServiceBase):
    async def Create(self, stream: Stream[CreateRequest, CreateResponse]) -> None:
        req = await stream.recv_message()
        assert req is not None

        blog = Blog(url=f"https://example.com/{req.username}")

        # do something

        res = CreateResponse(blog=blog)
        await stream.send_message(res)
